pallets/sto/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+
+            // TODO: Take custodial ownership of $sell_amount of $offering_token from primary issuance agent?
+            let fundraiser_id = Self::fundraiser_count(offering_asset) + 1;
+            let fundraiser = Fundraiser {
+                    offering_portfolio,
+                    offering_asset,
+                    raising_portfolio,
+                    raising_asset,
+                    tiers: tiers.into_iter().map(Into::into).collect(),
+                    venue_id,
+                    start: start.unwrap_or_else(Timestamp::<T>::get),
+                    end,
+                    frozen: false,
+            };
+            if let Some(end) = fundraiser.end {
+                ensure!(fundraiser.start < end, Error::<T>::FundraiserExpired);
+            }
+            FundraiserCount::insert(offering_asset, fundraiser_id);
+            <Fundraisers<T>>::insert(
+                offering_asset,
+                fundraiser_id,
+                fundraiser.clone()
+            );
+            Self::deposit_event(
+                RawEvent::FundraiserCreated(did, fundraiser)
+            );
+            Ok(())
+        }
+
+        /// Purchase tokens from an ongoing offering.
+        #[weight = 2_000_000_000]
+        pub fn invest(
+            origin,
+            investment_portfolio: PortfolioId,
+            funding_portfolio: PortfolioId,
+            offering_asset: Ticker,
+            fundraiser_id: u64,
+            investment_amount: T::Balance,
+            max_price: T::Balance,
+            reciept: Option<ReceiptDetails<T::AccountId, T::OffChainSignature>>
+        ) -> DispatchResult {
+            let sender = ensure_signed(origin.clone())?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+
+            ensure!(investment_portfolio.did == did, Error::<T>::InvalidPortfolio);
+            ensure!(funding_portfolio.did == did, Error::<T>::InvalidPortfolio);
+
+            let now = Timestamp::<T>::get();
+            let fundraiser = <Fundraisers<T>>::get(offering_asset, fundraiser_id).ok_or(Error::<T>::FundraiserNotFound)?;
+            ensure!(fundraiser.start <= now, Error::<T>::FundraiserNotStarted);
+            if let Some(end) = fundraiser.end {
+                ensure!(end > now, Error::<T>::FundraiserExpired);
+            }
+
+            // Remaining tokens to fulfil the investment amount
+            let mut remaining = investment_amount;
+            // Total cost to to fulfil the investment amount.
+            // Primary use is to calculate the blended price (offering_token_amount / cost).
+            // Blended price must be <= to max_price or the investment will fail.
+            let mut cost = T::Balance::from(0);
+            // Individual purchases from each tier that accumulate to fulfil the investment amount.
+            // Tuple of (tier_id, amount to purchase from that tier).
+            let mut purchases = Vec::new();
```

Comment:
It can be a with_capacity(tiers.length())

Reply:
it's not likely to span all tiers, it might be better to just let it naturally grow and not always alloc `tiers.length()`.